notebooks/visit_data.py

```diff
@@ -0,0 +1,255 @@
+
+# search mask, removes about 4%
+search_mask = visit_data_reduced['pagePath_parent'].str.contains('search')
+visit_data_reduced = visit_data_reduced[~search_mask]
+# empty mask, removes 3.4%
+empty_mask = visit_data_reduced['pagePath_parent'] == ''
+visit_data_reduced = visit_data_reduced[~empty_mask]
+# sign in mask, removes 3.8%
+sign_in_mask = visit_data_reduced['pagePath'].str.contains('sign-in')
+visit_data_reduced = visit_data_reduced[~sign_in_mask]
+# login mask, removes 0.2%
+login_mask = visit_data_reduced['pagePath'].str.contains('login')
+visit_data_reduced = visit_data_reduced[~login_mask]
+# log-in mask, removes 3.2%
+log_in_mask = visit_data_reduced['pagePath'].str.contains('log-in')
+visit_data_reduced = visit_data_reduced[~log_in_mask]
+# done mask, removes 1.0%
+done_mask = visit_data_reduced['pagePath_parent'] == 'done'
+visit_data_reduced = visit_data_reduced[~done_mask]
+
+fields_of_interest = ['fullVisitorId', 'visitId', 'pagePath', 'pagePath_abbreviated']
+visit_data_reduced[visit_data_reduced['fullVisitorId'] == 3250250007886851499][fields_of_interest]
+
+# lag1 page mask, removes 12%
+visit_data_reduced['pagePath_abbreviated_lag1'] = visit_data_reduced['pagePath_abbreviated'].shift()
+page_mask = visit_data_reduced['pagePath_abbreviated'] != visit_data_reduced['pagePath_abbreviated_lag1']
+visit_data_reduced = visit_data_reduced[page_mask]
+visit_data_reduced.shape
+# create a lag of visitor id to ensure that the lag1 page path relates to the same visitor
+# this cuts off the first stage of all journeys (as they have no valid lag 1 page path)
+# reduces the dataset down by a further 28%
+visit_data_reduced['fullVisitorId_lag1'] = visit_data_reduced['fullVisitorId'].shift()
+id_mask = visit_data_reduced['fullVisitorId'] == visit_data_reduced['fullVisitorId_lag1']
+visit_data_reduced = visit_data_reduced[id_mask]
+visit_data_reduced.shape
+visit_data_reduced[['pagePath_abbreviated_lag1', 'pagePath_abbreviated']]
+pd.DataFrame(visit_data_reduced['pagePath_abbreviated_lag1'].value_counts()).head(10)
+pd.DataFrame(visit_data_reduced['pagePath_abbreviated_lag1'].value_counts()).tail(10)
+
+# 34k levels of unique parent pagePath: might make it difficult to use knn, as its sparse
+len(visit_data_reduced['pagePath_abbreviated'].unique())
+
+# do a test train split and predict using the max count cat for the prior page to predict the subsequent page
+unique_visitorId = visit_data_reduced['fullVisitorId'].unique()
+random_df = pd.DataFrame({'fullVisitorId': unique_visitorId,
+                          'random_id': np.random.random(len(unique_visitorId))})
+visit_data_reduced = visit_data_reduced.merge(random_df, on='fullVisitorId')
+threshold_test_train = 0.7
+# train data
+fields_keep = ['date', 'fullVisitorId', 'visitId', 'hour', 'minute',
+               'pagePath_abbreviated', 'pagePath_abbreviated_lag1']
+train_data = visit_data_reduced[visit_data_reduced['random_id'] < threshold_test_train][fields_keep]
```

Comment:
Did you consider possible class imbalances? *e.g. If the random sample you took is 70% HMRC manuals, then the model will predict HMRC pages more frequently*

Reply:
This model relies upon that. If we balance up all the target classes, so they are evenly weighted, this will no longer work. Its effectively giving us the same accuracy as a resample of the underlying data. It could be an idea in practice to actually use random sampling from the actual distribution of pages, in order to create test data and explore the environment a bit.